apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -344,48 +448,128 @@ export async function aiGrade({
+        }
+        const GPTRubricGradeSchema = z.object({
+          rubric_items: GPTRubricItemSchema,
+          feedback: z.string(),
+        });
+        const completion = await openai.beta.chat.completions.parse({
+          messages,
+          model: 'gpt-4o-2024-08-06',
+          user: `course_${course.id}`,
+          response_format: zodResponseFormat(GPTRubricGradeSchema, 'grades'),
+        });
+        try {
+          msg += `Number of tokens used: ${completion.usage ? completion.usage.total_tokens : 0}\n`;
+          const grade_response = completion.choices[0].message;
+          msg += `Raw ChatGPT response:\n${grade_response.content}`;
+          if (grade_response.parsed) {
+            // Only care about the rubric numbers
+            const gpt_rubric_items: number[] = [];
+            for (let i = 0; i < rubric_items.length; i++) {
+              const item = grade_response.parsed.rubric_items[i];
+              if (item.selected) {
+                gpt_rubric_items.push(i);
+              }
+            }
+            const manual_rubric_data = {
+              rubric_id: rubric_items[0].rubric_id,
+              applied_rubric_items: parseRubricItems({
+                rubric_items,
+                gpt_rubric_items,
+              }),
+            };
+            new_rubric_items = await queryRows(
+              sql.select_rubric_for_grading,
+              {
+                assessment_question_id: assessment_question.id,
+              },
+              RubricItemSchema,
+            );
+            // Check if rubric items has been modified
+            validateRubric(rubric_items, new_rubric_items);
+            await manualGrading.updateInstanceQuestionScore(
+              assessment_question.assessment_id,
+              instance_question.id,
+              submission.id,
+              null, // modified_at
+              {
+                feedback: { manual: grade_response.parsed.feedback },
+                manual_rubric_data,
+              },
+              user_id,
+            );
+            msg += `\nAI rubric items: ${gpt_rubric_items.toString()}`;
+          } else if (grade_response.refusal) {
+            job.error(`ERROR AI grading for ${instance_question.id}`);
+            job.error(grade_response.refusal);
+            error_count++;
+          }
+        } catch (err) {
           job.error(`ERROR AI grading for ${instance_question.id}`);
-          job.error(grade_response.refusal);
+          job.error(err);
           error_count++;
         }
-      } catch (err) {
-        job.error(`ERROR AI grading for ${instance_question.id}`);
-        job.error(err);
-        error_count++;
+        job.info(msg);
+        if (warning) {
+          job.warn(warning);
+        }
+        rubric_items = new_rubric_items;
+      } else {
+        const completion = await openai.beta.chat.completions.parse({
+          messages,
+          model: 'gpt-4o-2024-08-06',
```

Comment:
We should have the model string only in one place. you could put at the top of the file

```js
OPEN_AI_MODEL = 'gpt-4o-2024-08-06'
```
and then use it in the places you need

Reply:
Or in config?

Reply:
I thought about that, the only reason I am hesitant to do that just yet is because we might want to use different models for different things at some point. So it doesn't necessarily make sense to have a global setting for it.

Reply:
I guess if the config item was `openAIGradingConfig` then it would be clear it shouldn't be used for everything

Reply:
For now I'm adding `OPEN_AI_MODEL = 'gpt-4o-2024-08-06'` to the ts file so we don't have to rely on someone updating the config file manually after some PR to update the model. It might also make customization in the future easier